File: javascript/bob/bob.js
const isShouting = s => /[A-Z]/.test(s) && s === s.toUpperCase()
const isQuestion = s => s.endsWith('?')
const isSilence = s => !/\S/.test(s)

module.exports = class Bob {
  hey(remark) {
    const s = String(remark)
    return isSilence(s) && "Fine. Be that way!"
      || isShouting(s) && "Whoa, chill out!"
      || isQuestion(s) && 'Sure.'
      || 'Whatever.'
  }
}

File: scala/bob/src/main/scala/Bob.scala

class Bob {
  def hey(remark: String): String = {
    if (remark.trim.isEmpty) "Fine. Be that way!"
    else if (isShouting(remark)) "Whoa, chill out!"
    else if (isQuestion(remark)) "Sure."
    else "Whatever."
  }

  // Shouting must contains one or more uppercase letters.
  private def isShouting(s: String): Boolean = {
    val letters = s.filter(_.isLetter)
    ! letters.isEmpty && letters.toUpperCase == letters
  }

  private def isQuestion(s: String): Boolean = s.endsWith("?")
}
